Search accessibility improvements, refs #13535

mInnes-archives opened the pull request:
### Description
PR seeks to improve search and advanced search form accessibility and compatibility with screen readers, as well as general clarity of function. This is achieved through the following:

- **Add aria-labels to search forms currently relying on placeholder text.** Not all screen reader use placeholder text.
- **Add ~~titles~~ aria-labels to search buttons.** This adds text for screen readers ~~and hover text for general clarity~~.
- **Associate form labels with existing forms.**
- **Change reset icon in advanced search to 'undo' for general clarity of function**. The 'x' has the same functionality as the reset action button at the bottom of advanced search. I felt that the 'undo' icon was a better representation for reset. 

For a fuller breakdown of the changes, see the commit descriptions.

Testing utilised Orca and NVDA screen readers.

### Images
Reset button icon changed from 'x' to 'undo'
![times_button](https://user-images.githubusercontent.com/76928773/121615426-d8d3d600-ca2e-11eb-9701-26788305bf09.png)
![undo_button](https://user-images.githubusercontent.com/76928773/121615429-da9d9980-ca2e-11eb-8a7f-0b43654a4f97.png)

mInnes-archives commented:
For reference, 2937d4d05946210de5098b8432c49163d14e021a was an attempt at addressing a routing bug with the "Exclude narrower terms" button in the "Related authority records" tab in term search. It currently routes back to the "Related Archival descriptions tab" rather than staying in the correct tab. I know what's causing it, but I just haven't created a fix yet. I also forgot that I started playing with this issue on this branch.

mInnes-archives commented:
I'm finding that using aria-labels instead of titles is considered better practice in general. The hover text of titles is only useful for mouse users and even then it's really just a crutch for bad UI design. Assuming icons are well selected, the titles are just excess.

mInnes-archives commented:
I added the original titles in d73bbd252acacc817a0e2c3b4362cc52600c1a29 in #1301. SInce the view options are also used for search, it's relevant to modify this here.

mInnes-archives commented:
fd011f3162daa735fa3d41648b3a4c2525b3e7a6  -> 'Region' and 'Region:' are treated as two seperate dictionary entries, so you need to separate out the punctuation for the translation to work. I'll keep an eye out for these as I work on different things. It should help cut down the size of the dictionary as well as redundant work when doing translations.

jraddaoui commented on apps/qubit/modules/actor/templates/_advancedSearch.php:
```diff
@@ -37,7 +37,7 @@
               </select>
 
-              <a href="#" class="delete-criterion"><i class="fa fa-times"></i></a>
+              <a href="#" class="delete-criterion" aria-label="<?php echo __('Delete criterion'); ?>"><i class="fa fa-times"></i></a>
```
Could you add `aria-hidden="true"` to the `i` element?

jraddaoui replied:
And if possible to all the others "affected" by this PR ...

This accessibility docs should still apply even if we're using a different Font Awesome version:

https://fontawesome.com/v5.15/how-to-use/on-the-web/other-topics/accessibility

jraddaoui commented on apps/qubit/modules/actor/templates/_advancedSearch.php:
```diff
@@ -26,7 +26,7 @@
               </select>
 
-              <input class="query" type="text" placeholder="<?php echo __('Search'); ?>" name="sq<?php echo $key; ?>" value="<?php echo $item['query']; ?>"/>
+              <input class="query" type="text" aria-label="<?php echo __('Search'); ?>" placeholder="<?php echo __('Search'); ?>" name="sq<?php echo $key; ?>" value="<?php echo $item['query']; ?>"/>
```
Not sure if `role="search"` would apply to this input (and all the others below). Feel free to ignore.

mInnes-archives replied:
Since `role="search"` is meant to be placed on the element containing all of the objects encompassing the search function, I've tried to implement this suggestion with this in mind (22942e8b633e09f681e0942652367f96bc8a02b8, 4b38262ce56e8b1789e4b5ece907b71bc754d57f).

I've created landmarks on both the inline search section and entirety of the advanced search section. I'm not sure if this is an appropriate approach, so I'd like to know your thoughts.

jraddaoui commented on apps/qubit/modules/repository/templates/_advancedFilters.php:
```diff
@@ -8,21 +8,8 @@
-    <div class="row-fluid">
-      <div class="span4">
-        <select name="thematicAreas">
+        <label for="thematicAreas"><?php echo __('Thematic area').':'; ?></label>
```
I'm not sure if formatting the translations this way may be an issue for RTL languages. Same apply for other translations below.

mInnes-archives replied:
That's a good point. I'm making an unjustified assumption on punctuation placement with this change. I'll revert the separation.

jraddaoui reviewed (changes requested):
This is really a great improvement @mInnes-archives!

I have added a few comments below and the same, check https://github.com/artefactual/atom/pull/1312#issuecomment-867393714 for the reference and how to rebase/merge this changes.

Thanks again for all these PRs, great additions!

jraddaoui commented on plugins/qtAccessionPlugin/modules/accession/templates/browseSuccess.php:
```diff
@@ -11,7 +11,8 @@
         <?php echo get_component('search', 'inlineSearch', [
-            'label' => __('Search %1%', ['%1%' => strtolower(sfConfig::get('app_ui_label_accession'))]), ]); ?>
+            'label' => __('Search accessions'),
+            'landmarkLabel' => __('Accession'), ]); ?>
```
This is something that has been puzzling me for a while and I'd like to get @fiver-watson input on it. There are a few places where we need to use the entities' UI labels in plural, like this one. Considering, i18n this is not something you can do manually so we end up not using those labels or using them in singular.

We could add another UI label setting for each entity to include a plural version. I'd consider it another issue if Dan likes the idea, but I'd keep using the UI labels in these cases until we address that one.

mInnes-archives replied:
Just to clarify, I removed the use of UI labels for 'accession' and 'donor' because, as far as I can tell, these labels were never actually created. I made a note of it and was thinking about adding them in a different PR. I also noticed the plural issue, but I can't immediately think of a better solution than what you've suggested.

fiver-watson replied:
Hey both, 

I do think that any changes around UI labels and plurality should probably be addressed in a separate PR/issue, so the work isn't lost in here, and/or blooms into something much larger that no longer fits the original PR purpose. 

This has certainly been an issue. In the past it's been dealt with on a case by case basis, so the solutions aren't always consistent across entity types. Mainly I think we need to establish a clear solution that can work for all the various page types and headers, and then focus on implementing it consistently going forward, rather than trying to sneak in a fix or two in different places, as this is part of what's led us to some inconsistency across entities. I'll add a few initial thoughts below, however. 

I do think there's a certain amount of mileage we can get from simply structuring strings a bit differently - for example, "Donor browse", "Authority record browse" etc could potentially work fine for most entities, but is not currently the norm. In some cases, adding "records" could solve a lot of issues around plurality (Browse donor records, browse repository records, browse archival description records, etc).... but it's a bit ungainly, and more importantly, it wouldn't work for authority records (Browse authority record records, lol).

Adding a whole set of separate label UI elements for plural cases would be nice, but also a fair bit of work, and likely needs consideration in terms of both ongoing maintenance, and impact on upgrading users (not sure how translations, fixtures, etc will factor into all of this for example). 

All this to say: I think it makes sense to address this as a separate issue, and inform any decision by trying to gather up some concrete examples of different places where we need to consider plurality.

jraddaoui commented on plugins/qtAccessionPlugin/modules/donor/templates/browseSuccess.php:
```diff
@@ -2,7 +2,7 @@
 
 <?php slot('title'); ?>
-  <h1><?php echo __('Browse donor'); ?></h1>
+  <h1><?php echo __('Browse donors'); ?></h1>
```
Another example, that wasn't even using the UI label, not even in singular.

jraddaoui reviewed (changes requested):
That looks really good to me @mInnes-archives!

I've added some comments below about the UI labels, nothing big and probably something we will want to address separately.

Thanks again, great improvements!

mInnes-archives committed:
Improve search accessibility, refs #13535

Squashed commit of the following:

commit d558d7936b43f01c2039113b9e8dea8d6e68ca0c
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jul 1 16:31:15 2021 -0400

    Add search landmark to move search

commit 1c977303474cc6cd8d1d34f1e5954f3ba4c88435
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jul 1 16:13:40 2021 -0400

    Add search landmark to hierarchy search

    - Use more precise label and placeholder text

commit 8a9fc2e7653b7061132f883dddcd2422953c698a
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jul 1 16:10:42 2021 -0400

    Improve accessibility of no-scoped holdings search

    - Add search landmark
    - Add aria-labels
    - Use app_ui_label_holdings

commit 94590101d19be524bc8df1d23f833f564effa678
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jul 1 15:22:39 2021 -0400

    Add search landmark to holdings search

commit 4268fb47a01029886221f48a854b2b91f288d219
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jul 1 14:46:45 2021 -0400

    Add search landmarks

    - Landmarks added to inline, advanced, and treeview searchs
    - Minor label grammatical number and casing changes
    - Remove use of non-existent donor and accession ui_labels

commit d3895384030be21e08a14efe3ff4937350c15929
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jun 24 11:28:45 2021 -0400

    Add aria-hidden attribute to icons

commit 46e351a226df8136b496f40379bf7b595164effe
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jun 24 11:11:00 2021 -0400

    Revert "Separate colon from translation"

    This reverts commit fd011f3162daa735fa3d41648b3a4c2525b3e7a6.

commit ec80bca45b61706f9444298e6055afd7231adb3c
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Jun 23 23:24:26 2021 -0400

    Separate colon from translation

commit 140805aff47bc2852d04e113a8d95f0e67f1ba18
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Jun 23 23:18:47 2021 -0400

    Normalize grammatical number on labels

commit 7d2499627f9feb798e0c20ec742d1d690a6a0897
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sat Jun 12 12:03:21 2021 -0400

    Change view option titles to aria-labels

commit 0905c84f94989970db79777cf1d332f97d4fdbd9
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sat Jun 12 11:52:23 2021 -0400

    Change title to aria-label

commit 1e43cdcc338d18606445519a473d6c4512d4fa6d
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sat Jun 12 11:42:15 2021 -0400

    Change titles to aria-labels

commit e1f3ce036a1701b21a6b0f3c106fea1c89d8a2ed
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Thu Jun 10 21:17:34 2021 -0400

    Revert unrelated code modification

    - Line pertains to an attempt at an another issue

commit 95422e89f3c29ab362191d0cca080d2d6d35a01c
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:47:09 2021 -0400

    Improve accessibility of term search

    - Add aria label to search form
    - Add titles to settings and search buttons
    - Add title to remove filter buttons

commit 30cee4b0db6ecf746c112a014b8323a7f28bd8c7
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:46:06 2021 -0400

    Improve accessibility of repository search

    - Associate form labels with forms
    - Add aria label to 'Search our collection' form
    - Add title to search button

commit fabd7f97ac2ebf23b350e63eb46c18676e468986
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:45:21 2021 -0400

    Improve accessibility of move search

    - Add aria label to search form
    - Add titles to search and reset buttons
    - Change icon for reset button to 'undo' for clarity
    - Add placeholder text for post-query search form

commit ecea8cf7cb8fad4403c128d188a01fa783b6c7ee
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:44:41 2021 -0400

    Improve accessibility of actor search

    - Add aria labels to advanced search forms
    - Add title to advanced search delete criterion button

commit 63f79b715d6c9aea6ec2ec0565778e894f1344a8
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:43:58 2021 -0400

    Improve accessibility of description search

    - Add aria label to treeview search form
    - Add title to search button
    - Add title to remove filter button for 'top-level' default filter
    - Add aria labels to advanced search forms
    - Add title to advanced search delete criterion button
    - Add title to help button

commit f65f24b2239675dce8f013ad22fbde2e4493c34c
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:42:27 2021 -0400

    Improve accessibilty of general search

    - Add title to remove filter button
    - Add aria label to search form
    - Add titles to search and reset buttons
    - Change icon for reset button to 'undo' for clarity
    - Add placeholder text for post-query search form

commit 34a7b577af795de470f07a9dc62ad9530243393c
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Tue Jun 8 23:41:05 2021 -0400

    Improve accessibilty of header search box

    - Add aria label to search form
    - Add title to search button
    - Remove unused text from search button

jraddaoui commented:
Thanks @fiver-watson and thanks again @mInnes-archives, great improvements!
